guides/05_chatbots/03_agents-and-tool-usage.md

```diff
@@ -346,3 +345,171 @@ This creates a chatbot that:
+                    "source": {
+                        "type": "base64",
+                        "media_type": "application/pdf",
+                        "data": pdf_data
+                    },
+                    "title": pdf_file.name,
+                    "citations": {"enabled": True}
+                })
+    
+    # Add the user's question
+    latest_message["content"].append({"type": "text", "text": history[-1]["content"]})
+    
+    formatted_messages.append(latest_message)
+    return formatted_messages
+```
+
+Then, let's create our bot response handler that processes citations:
+
+```python
+def bot_response(
+    history: list,
+    enable_citations: bool,
+    doc_type: str,
+    text_input: str,
+    pdf_file: str
+) -> List[Dict[str, Any]]:
+    try:
+        messages = format_message_history(history, enable_citations, doc_type, text_input, pdf_file)
+        response = client.messages.create(model="claude-3-5-sonnet-20241022", max_tokens=1024, messages=messages)
+        
+        # Initialize main response and citations
+        main_response = ""
+        citations = []
+        
+        # Process each content block
+        for block in response.content:
+            if block.type == "text":
+                main_response += block.text
+                if enable_citations and hasattr(block, 'citations') and block.citations:
+                    for citation in block.citations:
+                        if citation.cited_text not in citations:
+                            citations.append(citation.cited_text)
+        
+        # Add main response
+        history.append({"role": "assistant", "content": main_response})
+        
+        # Add citations in a collapsible section
+        if enable_citations and citations:
+            history.append({
+                "role": "assistant",
+                "content": "\n".join([f"• {cite}" for cite in citations]),
+                "metadata": {"title": "📚 Citations"}
+            })
+        
+        return history
+            
+    except Exception as e:
+        history.append({
+            "role": "assistant",
+            "content": "I apologize, but I encountered an error while processing your request."
+        })
+        return history
+```
+
+Finally, let's create the Gradio interface:
+
+```python
+with gr.Blocks() as demo:
+    gr.Markdown("# Chat with Citations")
+    
+    with gr.Row(scale=1):
+        with gr.Column(scale=4):
+            chatbot = gr.Chatbot(type="messages", bubble_full_width=False, show_label=False, scale=1)
+            msg = gr.Textbox(placeholder="Enter your message here...", show_label=False, container=False)
+            
+        with gr.Column(scale=1):
+            enable_citations = gr.Checkbox(label="Enable Citations", value=True, info="Toggle citation functionality" )
+            doc_type_radio = gr.Radio( choices=["plain_text", "pdf"], value="plain_text", label="Document Type", info="Choose the type of document to use")
```

Comment:
I agree that it would be better if users could upload multiple files and if `format_message_history` would be able to handle arbitrary mix of text and pdf files.

Other than that a including a gif would be nice!